### PR TITLE
Proof of concept for de-duplicating some coroutines code.

### PR DESCRIPTION
**Pull request by zach-klippenstein**
This pulls the common code from the `read` and `readAsync` methods out into an inline method that takes a function parameter for the actual read call. Inline functions can take suspending lambdas when called from suspend functions, so this is re-useable from both sync and async methods. This pattern could be copied for all the other methods.

### CODE REVIEW

**Comment by swankjesse** [comment]
Oooh, yes, this is a powerful technique. I think I can run with this.

File: okio/jvm/src/main/java/okio/RealBufferedSource.kt [hunk 1]
```diff
@@ -34,24 +34,25 @@ internal class RealBufferedSource(
...
-      if (read == -1L) return -1L
-    }
-
-    val toRead = minOf(byteCount, buffer.size)
-    return buffer.read(sink, toRead)
+    return readMaybeAsync(sink, byteCount, Source::read)
   }
 
   override suspend fun readAsync(sink: Buffer, byteCount: Long): Long {
+    // Can't use a suspend function reference here apparently.
```
Comment (Egorand) [inline okio/jvm/src/main/java/okio/RealBufferedSource.kt]: Hmm, is this a type inference problem, or a limitation of coroutines API?

Comment (zach-klippenstein) [reply okio/jvm/src/main/java/okio/RealBufferedSource.kt]: I think it's the former. The error message is something like:
```
Type mismatch: inferred type is KSuspendFunction0<Buffer, Long, Long> but (Buffer, Long) -> Long was expected
```

Comment (Egorand) [reply okio/jvm/src/main/java/okio/RealBufferedSource.kt]: Might be worth opening an issue on Kotlin's YouTrack.

Comment (zach-klippenstein) [reply okio/jvm/src/main/java/okio/RealBufferedSource.kt]: I just filed this issue: https://youtrack.jetbrains.net/issue/KT-28014

**Comment by swankjesse** [comment]
Abandoning this for now. Loom is a nicer solution!

**Pull request closed without merging**